### ISSUE TITLE
I have yolov5s.pt model and I would like to convert it to torchscript for android and with NNAPI support. It is not quite clear about how to convert the model to NNAPI supported model .ptl file

### ISSUE THREAD

**Issue by prashantenterprisesblp** [ultralytics/hub#377]
### Search before asking

- [X] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and [discussions](https://github.com/ultralytics/hub/discussions) and found no similar questions.

### Question

I have yolov5s.pt model and I would like to convert it to torchscript for android and with NNAPI support. It is not quite clear about how to convert the model to NNAPI supported model .ptl file.
I have converted my model successfully to .ptl model and it is running on android.
I would like to have better performance on mobile and would like to have NNAPI support.

**Comment by UltralyticsAssistant** [comment]
@prashantenterprisesblp hello, thanks for bringing up this issue!

Ultralytics Hub currently doesn't explicitly offer support for Android NNAPI with the torchscript models. NNAPI is meant for android hardware accelerators and TorchScript doesn't implicitly convert models to make full use of NNAPI. 

However, you can further optimize your .pt or .ptl model for mobile with PyTorch's optimization tools. PyTorch has a quantization process that you can apply to your model to better optimize for mobile, while keeping NNAPI requirements in mind. 

Remember to create a representative dataset that covers the operational domain of the model to get the most out of the quantization process. Also, be aware that quantization can often reduce the size and improve the speed of the model, but it might also reduce accuracy.

To use NNAPI, your model must comply with certain restrictions and requirements; not all PyTorch operations can be executed using NNAPI. So, you may need to adjust the architecture of your model accordingly.

Once you have optimized your model using PyTorch's tools, you should be able to deploy it on your android device.

These steps should help you achieve better performance on your Android application with NNAPI. Please let me know if you need any more help!

**Comment by prashantenterprisesblp** [comment]
Hi,
Thanks for the response. I have already optimized for mobile i.e. using
--optimize option while the export operation to .ptl file.
I would like to add further performance improvements using NNAPI.
Theoretically, I understand you, but I do not see any steps to do the same.
The sample to convert mobilenet model is having this line of code

model = torchvision.models.quantization.mobilenet.mobilenet_v2(pretrained=True,
quantize=quantize_core)

But, my model is not mobilenet and I do not see any option for yolov5s.

Do i need to create the .ptl file first using export.py and then use
the sample script on .ptl file?

The steps are are not clear.

**Comment by UltralyticsAssistant** [comment]
@prashantenterprisesblp hello,

Yes, that's correct. In addition to optimizing your models for mobile, applying the quantization process can potentially reduce the model size and improve the computational speed, which is highly beneficial for mobile deployments. 

However, it's crucial to remember that quantization might also lead to a decrease in model accuracy, so there's a trade-off to consider. It's also important to note that to leverage Android's NNAPI, your model has to comply with specific requirements and restrictions, which might require modifications to your model's architecture.

So in short: using the optimization option is good for mobile deployment but consider also using quantization for improved performance, while keeping in mind the potential reduction in accuracy.

Let me know if you have any further questions, I'm here to help!
